### DISCUSSION TITLE
When does the dokanloop thread exit

### DISCUSSION THREAD

**Discussion by zqlustc** [dokan-dev/dokany#976] [Q&A] [answered]
I have build the dokan environment. But I met the following problem.

thread A:
 -> call the mount

    thread B:
    mount thread start: ( It's the detached thread of thread A)
      ->fuse_loop_mt (it's a detached thread from my service)
        -> ...
          -> DokanMian
             -> dokanLoop(now is 10 child threads)
             -> WaitForMultipleObjects(will wait for child threads return, )
             -> ...... (release source)

 -> call the umount: (it's not ok)
  Now I umount the fuse through calling in main thread
    -> fuse_exit
    -> fuse_teardown
But the thread B is abruptly killed, and will not release source.

I want to know:
 (1) What might cause DokanMain(thread B) to terminate suddenly;
 (2) Under what circumstances will this dokanLoop threads end normally;
@Liryna

**Comment by Liryna** [comment, marked as answer]
fuse_exit will trigger an unmount which releases the thread B from the fuse_loop and call fuse_teardown that releases the dokan attached resources.
https://github.com/dokan-dev/dokany/blob/5a43c6c6f23d2214480629ef266d3731060634d9/dokan_fuse/src/dokanfuse.cpp#L810-L815

Thread B is never abruptly killed or terminate suddenly. It will leave the loop and follow the path described.

**Reply by zqlustc** [reply]
I  am reading the umount code in dokan,  a question troblems me.
```C
BOOL DOKANAPI DokanUnmount(WCHAR DriveLetter) {
  WCHAR mountPoint[MAX_PATH] = L"M:";
  mountPoint[0] = DriveLetter;
  return DokanRemoveMountPoint(mountPoint);
}
```

My mount path is like: `D:\DOKAN\DIR`，but the mountPoint passed to DokanRemoveMountPoint `D:`, is that right? And the umount always failed like:
```C
DokanError: Ioctl failed with code 122
Failed to unmount: D:
```
It confuses me

**Reply by Liryna** [reply]
Normally fuse_unmount will call DokanRemoveMountPoint and not DokanUnmount

**Reply by zqlustc** [reply]
I use the fuse_exit to trigger an unmount which releases the thread B from the fuse_loop.
```C
-> fuse_exit
  -> f->ch->ResolvedDokanUnmount(f->ch->mountpoint.at(0))
    -> DokanUnmount
      -> DokanRemoveMountPoint(mountPoint)
        -> SendGlobalReleaseIRP
error:
  DokanError: Ioctl failed with code 122
  Failed to unmount: D:
```
I want to release the thread B from the fuse_loop, But failed, but it always failed. fuse_exit will call the DokanUnmount firstly. My mount path is a dir `D:\DOKAN\DIR` in disk `D:`, not the `D:`

**Reply by Liryna** [reply]
That's a good point!
I made a change that should solve your case https://github.com/dokan-dev/dokany/commit/de4c04c449c046b103b1f64b6e3a554b26470020

Thank you for your feedback. Let me know if that fully solve your issue.

**Reply by zqlustc** [reply]
Too great, it could solve my problem with following changes together.
```C
@@ -785,7 +804,7 @@ void fuse_teardown(struct fuse *fuse, char *mountpoint) {
...
-  fuse_unmount(mountpoint, ch);
+ // fuse_unmount(mountpoint, ch);
...
 }
```
Thanks very much!!!
